Fix multiple MFA configurations login options bug

monfresh opened the pull request:
**Why**:
When a user has multiples of certain MFA configurations, we only want
to display the option once during login. For example, when a user
configures multiple Webauthn keys, either one will work, so we only
need to show the "Hardware security key" option once in the list of
security options displayed during login.

However, when a user has multiple MFA-enabled phone numbers, we want
to show them all so that the user can choose one in case they lost
access to one of them, for example.

The code in #2576 did not fix the issue with showing multiple
"Hardware security key" radio buttons. The feature specs added in this
PR fail on the master branch, and I confirmed it while testing locally.
I don't know why the presenter specs passed.

**How**:
- Use plain Ruby Enumerable methods to modify the presenters list so
that it only keeps one `WebauthnSelectionPresenter` instance, but allows
multiples of Sms and Voice.
- Remove unnecessary `WebauthnConfigurationsExtension` and related code.
- Update `PhoneSelectionPresenter#type` to use underscore instead of
colon to stay consistent with `id` names (Capybara doesn't like the
colon).
- Only show the last 4 digits of the phone number in the login options.

Hi! Before submitting your PR for review, and/or before merging it, please
go through the checklists below. These represent the more critical elements
of our code quality guidelines. The rest of the list can be found in
[CONTRIBUTING.md]

[CONTRIBUTING.md]: https://github.com/18F/identity-idp/blob/master/CONTRIBUTING.md#pull-request-guidelines

### Controllers

- [x] When adding a new controller that requires the user to be fully
authenticated, make sure to add `before_action :confirm_two_factor_authenticated`
as the first callback.

### Database

- [x] Unsafe migrations are implemented over several PRs and over several
deploys to avoid production errors. The [strong_migrations](https://github.com/ankane/strong_migrations#the-zero-downtime-way) gem
will warn you about unsafe migrations and has great step-by-step instructions
for various scenarios.

- [x] Indexes were added if necessary. This article provides a good overview
of [indexes in Rails](https://goo.gl/1DARYi).

- [x] Verified that the changes don't affect other apps (such as the dashboard)

- [x] When relevant, a rake task is created to populate the necessary DB columns
in the various environments right before deploying, taking into account the users
who might not have interacted with this column yet (such as users who have not
set a password yet)

- [x] Migrations against existing tables have been tested against a copy of the
production database. See #2127 for an example when a migration caused deployment
issues. In that case, all the migration did was add a new column and an index to
the Users table, which might seem innocuous.

### Encryption

- [x] The changes are compatible with data that was encrypted with the old code.

### Routes

- [x] GET requests are not vulnerable to CSRF attacks (i.e. they don't change
state or result in destructive behavior).

### Session

- [x] When adding user data to the session, use the `user_session` helper
instead of the `session` helper so the data does not persist beyond the user's
session.

### Testing

- [x] Tests added for this feature/bug
- [x] Prefer feature/integration specs over controller specs
- [x] When adding code that reads data, write tests for nil values, empty strings,
and invalid inputs.

jgsmith-usds commented:
#2576 should fix this.

monfresh commented:
@jgsmith-usds #2576 doesn't fix the issue. This PR does. PTAL. One spec is failing intermittently. I will update it later tonight. It needs to check the CSS based on the phone's configuration id, which is not always sms_1 and sms_2.

jgsmith-usds commented:
Why does it need to check the CSS? I'm not following something here. What's the bug you're trying to fix with this PR? I added a spec that did show that my fix worked for the case that shouldn't have shown multiple options (webauthn). We want to show all the phone options.

monfresh commented:
The bug I fixed is the same one that was supposedly fixed in #2576. As I mentioned in the commit message of this PR, the presenter spec in #2576 is misleading. Both feature specs in this PR (for multiple webauthn scenario, and multiple phones scenario) fail on the master branch. I confirmed this by testing locally as well. In #2567, if you have multiple webauthn keys, the hardware security key option doesn't show up at all in the login options. Same with when you have multiple phones.

In the case of multiple phones, the `id` of each sms/voice radio input element is of the form `two_factor_options_form_selection_{sms|voice}_1`, where the number at the end is the database ID of the phone configuration. In the feature spec, I was checking that each radio button only appears once for each phone number configured, and the spec was hardcoding the DB IDs to 1 and 2, but sometimes the IDs are different numbers. I just fixed the spec.

I also updated the login options to only show the last 4 digits of the phone number (I don't think an issue was opened for this but I remembered it while working on this PR).

monfresh commented on app/presenters/two_factor_authentication/phone_selection_presenter.rb:
```diff
@@ -2,18 +2,28 @@ module TwoFactorAuthentication
   class PhoneSelectionPresenter < SelectionPresenter
     def type
       if MfaContext.new(configuration&.user).phone_configurations.many?
-        "#{super}:#{configuration.id}"
```
As I mentioned in my commit message, it's for two reasons: consistency, and because colons are not valid CSS id selectors and causes Capybara to fail when looking for them. See the feature specs.

monfresh commented on app/presenters/two_factor_login_options_presenter.rb:
```diff
@@ -31,9 +31,15 @@ def options
+    # webauthn keys and phones. However, we only want to show one of each option
+    # during login, except for phones, where we want to allow the user to choose
+    # which MFA-enabled phone they want to use.
+    all_sms_and_voice_options_plus_one_of_each_remaining_option(options)
```
As I mentioned 3 times already, the other PR did not fix the multi-webauthn case. If you haven't already, please read this comment for details: https://github.com/18F/identity-idp/pull/2573#issuecomment-427533492

Also, this is less net code than we had before, by 5 lines.

jmhooper commented:
I confirmed the multi-webauthn bug locally. Here's what I see after logging in and choosing to use another option with 2 hardware keys setup:

![image](https://user-images.githubusercontent.com/963654/46693011-4a985980-cbce-11e8-88e8-9cd9897be2ee.png)
